cfpb/consumerfinance.gov
# TDP: Adds student financial capabilities surveys

This adds 3 web-based surveys about financial knowledge for grade school children ([PDF versions are here](https://www.consumerfinance.gov/consumer-tools/educator-tools/youth-financial-education/assess/)), as well as a system to embed resulting scores in a tamper-proof URL which can be shared with a teacher.

The shared URL also contains the student's initials obfuscated in a fragment (see `encode-name.js`) so that this info is never sent to the CFPB server nor captured by analytics. All survey temporary storage is based on client cookies, not DB/file sessions.

---

## Additions

- Mostly static survey intro pages for 3 grade ranges
- Multi-page survey system based on the Django `formtools` FormWizard component
- A python system to create a logical survey with pages of questions with all content loaded from hardcoded CSV
- Survey object methods to generate a custom form class for a single survey page
- A few Django form subclasses to customize markup within the dynamically created forms
- A python system to supply dynamic results (also from CSV) to a template based on user scores
- JS scripts to reduce data loss if a student refreshes/returns to a previous page
- JS scripts to encode/decode the user's initials, copy the shared URL to clipboard, and track progress
- JS scripts to reduce chances of a user accidentally leaking info to a later user

## How to test this PR

1. Navigate to a grade level page (see URLs below)
2. Enter the survey
3. Complete the survey
4. Choose to share the URL, enter initials, and copy the URL
5. The shared version requires no state and can be opened in an incognito window.

- https://cfgov-pr-6511.demo.cfpb.gov/consumer-tools/educator-tools/youth-financial-education/survey/3-5/
- https://cfgov-pr-6511.demo.cfpb.gov/consumer-tools/educator-tools/youth-financial-education/survey/6-8/
- https://cfgov-pr-6511.demo.cfpb.gov/consumer-tools/educator-tools/youth-financial-education/survey/9-12/

## Screenshots

![Screenshot from 2021-08-20 14-23-06](https://user-images.githubusercontent.com/170687/130277429-a54ce843-73b4-4ad8-a638-bcfddacdc49a.png)
![Screenshot from 2021-08-20 14-23-34](https://user-images.githubusercontent.com/170687/130277435-47547b55-7caa-4423-bdb4-2a35d81b9a72.png)
![Screenshot from 2021-08-20 14-24-07](https://user-images.githubusercontent.com/170687/130277441-e2637ec6-f621-49e6-8bf0-e6f287a61754.png)
![Screenshot from 2021-08-20 14-24-25](https://user-images.githubusercontent.com/170687/130277451-6b0e1199-575c-416d-a1d8-dbdd5c48716b.png)


## Notes and todos

- Backend tests will be added

## Checklist

  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

### Front-end testing

...

#### Browser testing

Aug 20 note: We're beginning more extensive testing this afternoon. We know that the use of the OS-default radio buttons results in, .e.g. red buttons in Firefox, however we have a plan to re-align the Django form markup with CFPB standards as soon as possible.

Visually tested in the following supported browsers:

#### Accessibility

Aug 0 note: I found a WAVE warning about duplicate radio button labels because of a Django form markup bug. We're hoping to address when re revisit the radio markup.


#### Other

The system is definitely dependent on JavaScript, and possibly CSS, for full operation.
